pep-0394.txt

```diff
@@ -5,104 +5,167 @@ Last-Modified: $Date$
+  version of Python available, and it can remove a dependency on Python 2.
+  The specifics are left on the distributors.
+  Examples of this include changing ``python`` shebangs to ``python3``
+  when Python 3.x is supported, changing ``python`` shebangs to ``python2``
+  when Python 3.x is not yet supported, or changing ``python3`` shebangs to
+  ``python3.8`` if the software is built with Python 3.8.
+* When a virtual environment (created by the PEP 405 ``venv`` package or a
+  similar tool such as ``virtualenv`` or ``conda``) is active, the ``python``
+  command should refer to the virtual environment's interpreter and should
+  always be available.
+  The ``python3`` or ``python2`` command (according to the environment's
+  interpreter) should also be available.
+
+For developers
+--------------
+
+* In order to tolerate differences across platforms, new code that needs
   to invoke the Python interpreter should not specify ``python``, but rather
-  should specify either ``python2`` or ``python3`` (or the more specific
-  ``python2.x`` and ``python3.x`` versions; see the `Migration Notes`_).
+  should specify either ``python3`` or ``python2`` (or the more specific
+  ``python3.x`` and ``python2.x`` versions; see the `Migration Notes`_).
   This distinction should be made in shebangs, when invoking from a shell
   script, when invoking via the system() call, or when invoking in any other
   context.
-* One exception to this is scripts that are deliberately written to be source
-  compatible with both Python 2.x and 3.x. Such scripts may continue to use
-  ``python`` on their shebang line.
-* When packaging software that is source compatible with both versions,
-  distributions may change such ``python`` shebangs to ``python3``.
-  This ensures software is used with the latest version of
-  Python available, and it can remove a dependency on Python 2.
 * When reinvoking the interpreter from a Python script, querying
   ``sys.executable`` to avoid hardcoded assumptions regarding the
   interpreter location remains the preferred approach.
-* In controlled environments aimed at expert users, where being explicit
-  is valued over user experience (for example, in test environments and
-  package build systems), distributions may choose to not provide the
-  ``python`` command even if ``python2`` is available.
-  (All software in such a controlled environment must use ``python3`` or
-  ``python2`` rather than ``python``, which means scripts that deliberately
-  use ``python`` need to be modified for such environments.)
-* When a virtual environment (created by the PEP 405 ``venv`` package or a
-  similar tool) is active, the ``python`` command should refer to the
-  virtual environment's interpreter. In other words, activating a virtual
-  environment counts as deliberate user action to change the default
-  ``python`` interpreter.
+* Scripts that are deliberately written to be source compatible with both
```

Comment:
I think we need to present additional options here, as our actual advice to developers is "don't rely on hardcoded shebang lines, and instead let tools rewrite them for you".

That means the preferred option is "Use a Python package installer that supports shebang line rewriting, like `pip install --user` or `pipx install`, rather than copying around plain text files" (perhaps with a reference to https://packaging.python.org/guides/installing-stand-alone-command-line-tools/ ).

Option 2 is "Use the environment-aware `/usr/bin/env python` shebang line, and require that the script be executed inside an activated Python virtual environment"

Option 3 is "Use the generic `/usr/bin/env python` shebang line, but ensure that will resolve to a suitable interpreter on the systems that you're targeting if the script needs to run outside a virtual environment". Adding in `ln -s ~/.local/bin/python /usr/bin/python3` or `sudo ln -s /usr/local/bin/python /usr/bin/python3` for the target systems that you care about is a much lower impact change than rewriting all your shebang lines to add a version qualifier (especially since making that change will break the scripts on systems that only have Python 2).

The only time that using the `python3` shebang genuinely makes sense is when the script wouldn't run on Python 2 anyway - in those cases, using the more specific shebang only increases your target platform compatibility, it never reduces it.

Reply:
If we don't want that level of complexity in the PEP, than I think we should make the guidance explicitly:

- `/usr/bin/env python` for source compatible scripts, or Python 3 scripts that are unlikely to ever be exposed to target environments where that invocation won't work (i.e. Python 3 only systems when outside a virtual environment)
- `/usr/bin/env python3` for scripts that require Python 3 or later
- `/usr/bin/env python2` for scripts that don't support Python 3 at all
- where possible, use a Python package or script installer (e.g. `pip` or `pipx`) instead of copying scripts directly to a target system, as they are able to rewrite shebang lines to suit the target environment

Hardcoding specific paths to Python binaries should be restricted to non-portable scripts that only target particular platforms and don't need to respect virtual environments.

Reply:
After discussing this with @encukou in person, I created a PR against the PR branch to better convey what I'm proposing here: https://github.com/hroncok/peps/pull/1/files